sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -87,6 +99,116 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+
+            public AccessTokenCache(TimeSpan tokenRefreshOffset)
+            {
+                _tokenRefreshOffset = tokenRefreshOffset;
+            }
+
+            public string? SaveToken(AccessToken token, Exception? exception)
+            {
+                string? headerValue;
+                TaskCompletionSource<string?>? pendingTcs;
+
+                lock (_syncObj)
+                {
+                    if (!string.IsNullOrEmpty(token.Token))
+                    {
+                        _headerValue = "Bearer " + token.Token;
+                        _refreshOn = token.ExpiresOn - _tokenRefreshOffset;
+                        _expiresOn = token.ExpiresOn;
+                        _tokenState = TokenState.Valid;
+                    }
+                    else
+                    {
+                        _headerValue = default;
+                        _refreshOn = default;
+                        _expiresOn = default;
+                        _tokenState = TokenState.Invalid;
+                    }
+
+                    pendingTcs = _pendingTcs;
+                    headerValue = _headerValue;
+                    _pendingTcs = null;
+                }
+
+                if (pendingTcs == null)
+                {
+                    return headerValue;
+                }
+
+                if (headerValue != null)
+                {
+                    pendingTcs.SetResult(headerValue);
+                }
+                else
+                {
+                    pendingTcs.SetException(exception ?? new InvalidOperationException($"{nameof(TokenCredential)}.{nameof(TokenCredential.GetToken)} has failed with unknown error."));
+                }
+
+                return headerValue;
+            }
+
+            public ValueTask<string?> GetTokenAsync()
+            {
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn && _tokenState != TokenState.Pending)
+                    {
+                        _tokenState = TokenState.Pending;
+                        _headerValue = null;
+                        return new ValueTask<string?>();
+                    }
+
+                    if (DateTimeOffset.UtcNow >= _refreshOn && _tokenState == TokenState.Valid)
+                    {
+                        _tokenState = TokenState.AboutToExpire;
+                        return new ValueTask<string?>();
+                    }
+
+                    switch (_tokenState) {
+                        case TokenState.Pending:
+                            _pendingTcs ??= new TaskCompletionSource<string?>(TaskCreationOptions.RunContinuationsAsynchronously);
+                            return new ValueTask<string?>(_pendingTcs.Task);
+                        case TokenState.Valid:
+                        case TokenState.AboutToExpire:
+                            return new ValueTask<string?>(_headerValue);
+                        default:
```

Comment:
What about TokenState.Invalid?

Reply:
Handled above by:
```cs
if (DateTimeOffset.UtcNow >= _expiresOn && _tokenState != TokenState.Pending)
{
    _tokenState = TokenState.Pending;
    _headerValue = null;
    return new ValueTask<string?>();
}
```

If state is `Invalid` and `DateTimeOffset.UtcNow < _expiresOn` then something went wrong.